packages/core/utils/src/env-helper.ts

```diff
@@ -10,90 +13,130 @@ function getKey(key: string) {
-    if (!_.has(process.env, key)) {
-      return defaultValue;
+function envJson(key: string): object | undefined;
+function envJson(key: string, defaultValue: object): object;
+function envJson(key: string, defaultValue: undefined): object | undefined;
+function envJson(key: string, defaultValue?: object): object | undefined {
+  if (!_.has(process.env, key)) {
+    return defaultValue;
+  }
+
+  try {
+    return JSON.parse(getKey(key));
+  } catch (error) {
+    if (error instanceof Error) {
+      throw new Error(`Invalid JSON environment variable ${key}: ${error.message}`);
     }
 
-    let value = getKey(key);
-
-    if (value.startsWith('[') && value.endsWith(']')) {
-      value = value.substring(1, value.length - 1);
-    }
+    throw error;
+  }
+}
 
-    return value.split(',').map((v) => {
-      return _.trim(_.trim(v, ' '), '"');
-    });
-  },
+function envArray(key: string): ReadonlyArray<string> | undefined;
+function envArray(key: string, defaultValue: ReadonlyArray<string>): ReadonlyArray<string>;
+function envArray(key: string, defaultValue: undefined): ReadonlyArray<string> | undefined;
+function envArray(
+  key: string,
+  defaultValue?: ReadonlyArray<string>
+): ReadonlyArray<string> | undefined {
+  if (!_.has(process.env, key)) {
+    return defaultValue;
+  }
+
+  let value = getKey(key);
+
+  if (value.startsWith('[') && value.endsWith(']')) {
+    value = value.substring(1, value.length - 1);
+  }
+
+  return value.split(',').map((v) => {
+    return _.trim(_.trim(v, ' '), '"');
+  });
+}
 
-  date(key: string, defaultValue?: Date): Date | undefined {
-    if (!_.has(process.env, key)) {
-      return defaultValue;
-    }
+function envDate(key: string): Date | undefined;
+function envDate(key: string, defaultValue: Date): Date;
+function envDate(key: string, defaultValue: undefined): Date | undefined;
+function envDate(key: string, defaultValue?: Date): Date | undefined {
+  if (!_.has(process.env, key)) {
+    return defaultValue;
+  }
 
-    return new Date(getKey(key));
-  },
-
-  /**
-   * Gets a value from env that matches oneOf provided values
-   * @param {string} key
-   * @param {string[]} expectedValues
-   * @param {string|undefined} defaultValue
-   * @returns {string|undefined}
-   */
-  oneOf(key: string, expectedValues?: unknown[], defaultValue?: unknown) {
-    if (!expectedValues) {
-      throw new Error(`env.oneOf requires expectedValues`);
-    }
+  return new Date(getKey(key));
+}
 
-    if (defaultValue && !expectedValues.includes(defaultValue)) {
-      throw new Error(`env.oneOf requires defaultValue to be included in expectedValues`);
-    }
+/**
+ * Gets a value from env that matches oneOf provided values
+ */
+function envOneOf<T extends string>(key: string, expectedValues: ReadonlyArray<T>): T | undefined;
+function envOneOf<T extends string>(
+  key: string,
+  expectedValues: ReadonlyArray<T>,
+  defaultValue: NoInfer<T>
```

Comment:
Note: `NoInfer` is a `typescript@5.4` feature.

https://www.typescriptlang.org/docs/handbook/release-notes/typescript-5-4.html#the-noinfer-utility-type